src/monodroid/jni/monodroid-glue.c

```diff
@@ -704,55 +737,192 @@ typedef void* (*mono_mkbundle_init_ptr) (void (*)(const MonoBundledAssembly **),
-	if (!mono_mkbundle_initialize_mono_api)
-		log_error (LOG_BUNDLE, "Missing initialize_mono_api in the application");
+	void *handle = dlopen (path, dl_flags);
+	if (handle == NULL)
+		log_info (LOG_ASSEMBLY, "Failed to load shared library '%s'. %s", path, dlerror ());
+	return handle;
+}
 
-	mono_mkbundle_init = dlsym (libapp, "mono_mkbundle_init");
-	if (!mono_mkbundle_init)
-		log_error (LOG_BUNDLE, "Missing mono_mkbundle_init in the application");
-	log_info (LOG_BUNDLE, "Bundled app loaded: %s", libappso);
+static void*
+load_dso_from_app_lib_dirs (const char *name, int dl_flags)
+{
+	if (name == NULL)
+		return NULL;
+
+	mono_bool needs_free = FALSE;
+	char *full_path = NULL;
+	for (int i = 0; i < app_lib_directories_size; i++) {
+		full_path = get_full_dso_path (app_lib_directories [i], name, &needs_free);
+		void *handle = load_dso (full_path, dl_flags, FALSE);
+		if (needs_free)
+			free (full_path);
+		if (handle != NULL)
+			return handle;
+	}
+
+	return NULL;
+}
+
+static void*
+load_dso_from_override_dirs (const char *name, int dl_flags)
+{
+	if (name == NULL)
+		return NULL;
+
+	for (int i = 0; i < MAX_OVERRIDES; i++) {
+		if (override_dirs [i] == NULL)
+			continue;
+
+		mono_bool needs_free = FALSE;
+		char *full_name = get_full_dso_path (override_dirs [i], name, &needs_free);
+		void *handle = load_dso (full_name, dl_flags, FALSE);
+		if (needs_free)
+			free (full_name);
+		if (handle != NULL)
+			return handle;
+	}
+
+	return NULL;
 }
 
 static char*
-get_bundled_app (JNIEnv *env, jstring dir)
+get_existing_dso_path_on_disk (const char *base_dir, const char *dso_name, mono_bool *needs_free)
 {
-	const char *v;
-	char *libapp;
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
 
-#ifndef RELEASE
-	libapp = path_combine (override_dirs [0], "libmonodroid_bundle_app.so");
+	*needs_free = FALSE;
+	char *dso_path = get_full_dso_path (base_dir, dso_name, needs_free);
+	if (file_exists (dso_path))
+		return dso_path;
 
-	if (file_exists (libapp))
-		return libapp;
+	*needs_free = FALSE;
+	free (dso_path);
+	return NULL;
+}
 
-	free (libapp);
-#endif
+static void
+__dso_alloc_cleanup (char **dso_path, mono_bool *needs_free)
+{
+	if (dso_path != NULL) {
+		if (needs_free && *needs_free)
+			free (*dso_path);
+		*dso_path = NULL;
+	}
 
-	if (dir) {
-		v = (*env)->GetStringUTFChars (env, dir, NULL);
-		if (v) {
-			libapp = path_combine (v, "libmonodroid_bundle_app.so");
-			(*env)->ReleaseStringUTFChars (env, dir, v);
-			if (file_exists (libapp))
-				return libapp;
-		}
+	if (needs_free)
+		*needs_free = FALSE;
+}
+
+static char*
+get_full_dso_path_on_disk (const char *dso_name, mono_bool *needs_free)
+{
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
+
+	*needs_free = FALSE;
+	if (embedded_dso_mode)
+		return NULL;
+
+	char *dso_path = NULL;
+	for (int i = 0; i < MAX_OVERRIDES; i++) {
```

Comment:
`override_dirs` should only be checked in non`-RELEASE` builds.

Reply:
it can be but it's a no-op because `override_dirs` aren't filled up in Release mode. I like to avoid unnecessary `#ifdef`s...

Reply:
Admittedly, `#if`s are ugly.  However, they're also easier to audit.  Without the `#if`, it means we need to consider runtime behavior, and if we accidentally alter runtime behavior such that `override_dirs` is filled in non-`RELEASE` builds, we now have a security vulnerability.

Take, for example, `create_update_dir()`: https://github.com/xamarin/xamarin-android/blob/b1d81460628e8cf2d78439ae760ef6b065db3a27/src/monodroid/jni/monodroid-glue.c#L423-L442

This is [always invoked](https://github.com/xamarin/xamarin-android/blob/b1d81460628e8cf2d78439ae760ef6b065db3a27/src/monodroid/jni/monodroid-glue.c#L3785), and in `RELEASE` builds will create a `.__override__` directory when the `debug.mono.profile` system property is set.

Which means that if we *don't* protect `override_dirs` here, if the `debug.mono.profile` system property is set, we have a security vulnerability.

Yes, `#if` is ugly .  It also makes *auditing* easier (don't need to worry about/track down runtime behaviors) and maintenance easier (less likely to break things accidentally if it cannot possibly happen in the first place).